### DISCUSSION TITLE
UI Testing w\ ReactiveUI, Splat, and WhenActivated

### DISCUSSION THREAD

**Discussion by trippsc2** [AvaloniaUI/Avalonia#13804] [Q&A] [answered]
Hello.

I have a project that I'm working on using Avalonia with ReactiveUI.  I've been trying to use the `Avalonia.Headless.XUnit` package to do UI tests, but I've been running into problems.

- If I resolve my MainViewModel like I do in the app using `Locator.Current.GetService`, I run into issues of shared state between tests that cannot easily be reset.
- If I create my own `ModernDependencyResolver` instance, manage its lifecycle on a per test collection basis or on a per test basis, and use it to resolve my `MainViewModel`, I run into threading issues that throw `InvalidOperationException` regarding the object state.

I do most of my bindings and subscriptions within `WhenActivated` blocks.  I haven't concluded if that is related or not, but that doesn't seem to be the norm in Avalonia projects I've seen that use ReactiveUI.

I have test parallelization disabled assembly-wide in the test project to ensure that was not causing issues.

Does anyone have any ideas what I might be running into and/or know of any projects I might review that have this kind of testing working?

**Comment by thevortexcloud** [comment]
I am not sure what your tests are structured like and I can't say I have written a UI test with Avalonia before, but I imagine the first point could be solved by creating a new DI container/app instance per test. Typically with UI tests you run a specific sequence, then you either tear the entire thing down and start again; or you run your tests sequentially and return to a known clean state to guarantee state consistency. This is specifically avoid side effects like this. iOS for example apparently will run separate simulated devices if you try to execute tests in parallel.

**Comment by trippsc2** [comment]
My tests are currently inheriting from the following.

    public abstract class ViewTests : IDisposable
    {
        protected readonly CompositeDisposable Disposables = new();
        
        protected IDependencyResolver CreateResolver()
        {
            var resolver = new ModernDependencyResolver();
            resolver.RegisterConstant(
                Locator.Current.GetService<IActivationForViewFetcher>(),
                typeof(IActivationForViewFetcher));
            resolver.RegisterConstant(
                Locator.Current.GetService<IPropertyBindingHook>(),
                typeof(IPropertyBindingHook));
            
            resolver.RegisterOpenTrackerNextCoreTypes();
            resolver.RegisterOpenTrackerNextEditorTypes();

            resolver.DisposeWith(Disposables);
            
            return resolver;
        }

        protected MainWindow CreateMainWindow(IDependencyResolver resolver)
        {
            var mainWindow = new MainWindow(
                resolver.GetService<IImageService>(),
                resolver.GetService<IServiceCollection<IPackFolderService>>())
            {
                DataContext = resolver.GetService<MainViewModel>()
            };
            
            mainWindow.Show();
            Dispatcher.UIThread.RunJobs();

            Disposable
                .Create(() =>
                {
                    mainWindow.Close();
                    Dispatcher.UIThread.RunJobs();
                })
                .DisposeWith(Disposables);
            
            return mainWindow;
        }

        public void Dispose()
        {
            Disposables.Dispose();
            
            GC.SuppressFinalize(this);
        }
    }

I'm calling the `CreateResolver()` method and `CreateMainWindow(IDependencyResolver)` method at the beginning of the tests.  The tests have the `[AvaloniaFact]` attribute on them.  My AppBuilder code is the following.

    [assembly: AvaloniaTestApplication(typeof(TestAppBuilder))]
    [assembly: CollectionBehavior(DisableTestParallelization = true)]

    namespace OpenTrackerNext.Editor.UnitTests;

    [ExcludeFromCodeCoverage]
    public sealed class TestAppBuilder
    {
        public static AppBuilder BuildAvaloniaApp()
        {
            IconProvider.Current.Register<MaterialDesignIconProvider>();
            
            return AppBuilder
                .Configure<App>()
                .UseReactiveUI()
                .UseHeadless(new AvaloniaHeadlessPlatformOptions());
        }
    }

When I run the tests in Rider in capturing coverage or without debugging, the tests all complete without exception, but no changes are captured after the initial state.  When I run the tests with debugging tools, I get an `InvalidOperationException` when running `Dispatcher.UIThread.RunJobs()` for the first time.  If I omit that call, no changes from bindings propagate to the view.

**Reply by trippsc2** [reply, marked as answer]
I found a DataTemplate that was referencing `Locator.Current` that I believe was causing the issue.  I ended up changing my setup methods to the following and the issue appears to be resolved.

```c#
public abstract class ViewTests : IDisposable
{
    protected readonly CompositeDisposable Disposables = new();

    protected MainWindow MainWindow { get; private set; } = default!;
    
    public void Dispose()
    {
        Disposables.Dispose();
        
        GC.SuppressFinalize(this);
    }
    
    protected void CreateResolver()
    {
        var resolver = new ModernDependencyResolver();

        resolver.RegisterConstant(
            Locator.Current.GetService<IActivationForViewFetcher>(),
            typeof(IActivationForViewFetcher));
        resolver.RegisterConstant(
            Locator.Current.GetService<IPropertyBindingHook>(),
            typeof(IPropertyBindingHook));
        
        resolver.RegisterOpenTrackerNextCoreTypes();
        resolver.RegisterOpenTrackerNextEditorTypes();

        var previousResolver = Locator.GetLocator();

        Locator.SetLocator(resolver);

        Disposable.Create(() =>
            {
                Locator.SetLocator(previousResolver);
                resolver.Dispose();
            })
            .DisposeWith(Disposables);
    }

    protected void CreateMainWindow()
    {
        MainWindow = new MainWindow
        {
            DataContext = Locator.Current.GetService<MainViewModel>()
        };
        
        MainWindow.Show();
        Dispatcher.UIThread.RunJobs();

        Disposable
            .Create(() =>
            {
                MainWindow.Close();
                Dispatcher.UIThread.RunJobs();
            })
            .DisposeWith(Disposables);
    }
}
```
